### DISCUSSION TITLE
Fabric.js : Cannot achieve Text with Patern in 6.0.2

### DISCUSSION THREAD

**Discussion by predragslavkovic** [fabricjs/fabric.js#9981] [Q&A] [answered]
Hey,

I'm using fabric.js version 6.0.2.

I'm trying to add pattern on text but apparently I am doing something wrong since I am new to Fabric.js and it's hard for me to wrap my mind around it. Before I am writing here I was trying to find everywhere how to achieve this in 6.0.x but unsuccessfully. The Demo that I've used for reference is one from here: http://fabricjs.com/patterns
This is my code:

`
    var canvas = new fabric.Canvas('lava-design-canvas');
    window.lavaDesignCanvas = canvas;
    
    function loadPattern(url) {
        console.log(url);
        fabric.util.loadImage(url, function(img) {
            text.set('fill', new fabric.Pattern({
                source: img,
                repeat: 'repeat'
            }));
            shape.set('fill', new fabric.Pattern({
                source: img,
                repeat: 'repeat'
            }));
            canvas.renderAll();
        });
    }

    var text = new fabric.Text('Honey,\nI\'m subtle', {
        fontSize: 250,
        left: 0,
        top: 0,
        lineHeight: 1,
        originX: 'left',
        fontFamily: 'Helvetica',
        fontWeight: 'bold',
        statefullCache: true,
        scaleX: 0.4,
        scaleY: 0.4
    });
    var shape = new fabric.Rect({
        width: 200,
        height: 100,
        left: 10,
        top: 300,
    });
    canvas.add(text, shape);

    loadPattern('https://lavadesign.rs/wp-content/uploads/2024/07/gold_grad.png');
`

Any idea where I am making mistake ?

**Comment by asturur** [comment]
in fabric 6 all the functions that take a callback return now a promise. Is described here: https://github.com/fabricjs/fabric.js/issues/8299
Are you using typescript or vscode? that will suggest you a lot of the api now

**Reply by predragslavkovic** [reply]
Thank you @asturur for your answer. Before I realize that the functions return a promise now, I tried to work around this and found this solution where instead of load the image asynchronously I took an image from the DOM and it worked fine.

```
const pattern_gold = new fabric.Pattern({
    source: document.getElementById('svg_gold'),
});

goldPrintButton.addEventListener('click', function () {
    var activeObject = canvas.getActiveObject();
    if (activeObject && activeObject.type === 'textbox') {
        activeObject.set('fill', pattern_gold);
        canvas.renderAll();
    }
});
```
    
 I am writing vanilla JavaScript in PhpStorm :/   
Thank you for your answer, appreciate it 🙏

**Reply by asturur** [reply, marked as answer]
your old code would work with:
```
fabric.util.loadImage(url).then((img) => {
        text.set('fill', new fabric.Pattern({
            source: img,
            repeat: 'repeat'
        }));
        shape.set('fill', new fabric.Pattern({
            source: img,
            repeat: 'repeat'
        }));
        canvas.requestRenderAll();
    });
```

**Reply by asturur** [reply]
Check if there is a typescript plugin or a jsdocs plugin for php storm, it helps a lot with suggestions and static analisys

**Reply by predragslavkovic** [reply]
I'll switch to some JS dedicated IDE for sure for this kind of stuff.

Thanks once again!
